src/zino/tasks/bgpstatemonitortask.py

```diff
@@ -0,0 +1,404 @@
+    "cisco": ("CISCO-BGP4-MIB", "cbgpLocalAs"),
+    "general": ("BGP4-MIB", "bgpLocalAs"),
+}
+BUGGY_REMOTE_ADDRESSES = [
+    # Bug in JunOS -- info from IPv6 BGP sessions spill over
+    "0.0.0.0",
+    # Bug in earlier Cisco IOS, info from elsewhere (IPv6?) spills over
+    "32.1.7.0",
+]
+
+
+@dataclass
+class BaseBGPRow:
+    peer_state: BGPOperStates
+    peer_admin_status: BGPAdminStatus
+    peer_remote_address: IPAddress
+    peer_remote_as: int
+    peer_fsm_established_time: int
+
+
+class BGPStateMonitorTask(Task):
+    """Fetches and stores state information about external BGP sessions."""
+
+    async def run(self):
+        bgp_style = await self._get_bgp_style()
+        if bgp_style != self.device_state.bgp_style:
+            _logger.debug(
+                f"Router {self.device_state.name} changed its BGP style from '{self.device_state.bgp_style}' to "
+                f"'{bgp_style}'"
+            )
+            self.device_state.bgp_style = bgp_style
+        if not bgp_style:
+            return
+
+        local_as = await self._get_local_as(bgp_style=bgp_style)
+        uptime = await self._get_uptime()
+        if not local_as or not uptime:
+            return
+
+        if bgp_style == BGPStyle.JUNIPER:
+            bgp_info = await self._get_juniper_bgp_info()
+        elif bgp_style == BGPStyle.CISCO:
+            bgp_info = await self._get_cisco_bgp_info()
+        elif bgp_style == BGPStyle.GENERAL:
+            bgp_info = await self._get_general_bgp_info()
+
+        if not bgp_info:
+            return
+
+        bgp_info = self._fixup_ip_addresses(bgp_info=bgp_info)
+
+        # BGP peers is by default None, so to record the BGP peers, we have to initialize it
+        if not self.device_state.bgp_peers:
+            self.device_state.bgp_peers = dict()
+
+        for result in bgp_info.values():
+            self._update_single_bgp_entry(row=result, local_as=local_as, uptime=uptime)
+
+    async def _get_bgp_style(self) -> Optional[BGPStyle]:
+        if await self.snmp.subtree_is_supported("BGP4-V2-MIB-JUNIPER", "jnxBgpM2"):
+            return BGPStyle.JUNIPER
+
+        if await self.snmp.subtree_is_supported("CISCO-BGP4-MIB", "cbgpPeer2Table"):
+            return BGPStyle.CISCO
+
+        if await self.snmp.subtree_is_supported("BGP4-MIB", "bgp"):
+            return BGPStyle.GENERAL
+
+        return None
+
+    async def _get_local_as(self, bgp_style: BGPStyle) -> Optional[int]:
+        try:
+            mib, object_name = LOCAL_AS_OBJECTS[bgp_style]
+        except KeyError:
+            return
+        response = await self.snmp.getnext(mib, object_name)
+        if self.snmp.is_in_scope(response, (mib, object_name)):
+            return response.value
+        else:
+            _logger.info(f"router {self.device.name} misses {object_name}")
+
+    async def _get_uptime(self) -> Optional[int]:
+        try:
+            return (await self.snmp.get("SNMPv2-MIB", "sysUpTime", 0)).value / 100
```

Comment:
The parent Task class has a built-in _get_uptime method that can be used (dont know if that this PR has that update uncluded)